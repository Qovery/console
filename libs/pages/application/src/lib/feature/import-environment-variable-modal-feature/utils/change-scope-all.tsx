import { APIVariableScopeEnum } from 'qovery-typescript-axios'
import { UseFormSetValue } from 'react-hook-form'

export function changeScopeForAll(
  value: APIVariableScopeEnum,
  setValue: UseFormSetValue<{ [key: string]: string }>,
  keys: (string | boolean)[]
): void {
  keys.forEach((key) => {
    setValue(key + '_scope', value?.toString())
  })
}
