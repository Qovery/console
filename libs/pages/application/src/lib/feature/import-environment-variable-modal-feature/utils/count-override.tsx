export function countOverride(data: { [key: string]: string }, existingEnvVarNames: (string | undefined)[]) {
  let count = 0
  Object.keys(data)
    .filter((key) => key.indexOf('_key') >= 0)
    .forEach((key) => {
      if (existingEnvVarNames.includes(data[key])) count++
    })

  return count
}
