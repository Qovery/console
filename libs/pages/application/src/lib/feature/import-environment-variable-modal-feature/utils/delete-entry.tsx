export function deleteEntry(
  key: string,
  setKeys: (keys: string[]) => void,
  keys: string[],
  unregister: (name: string | string[]) => void
): void {
  const newKeys = keys.filter((el) => el !== key)
  setKeys(newKeys)
  unregister(key + '_key')
  unregister(key + '_value')
  unregister(key + '_secret')
  unregister(key + '_scope')
}
