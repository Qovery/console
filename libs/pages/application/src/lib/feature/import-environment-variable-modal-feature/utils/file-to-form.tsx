import { APIVariableScopeEnum } from 'qovery-typescript-axios'

export function jsonToForm(json: string): { [key: string]: string } {
  const parsed = JSON.parse(json)
  return parsedToForm(parsed)
}

export function parsedToForm(parsed: { [key: string]: string }): { [key: string]: string } {
  const defaultValues: { [key: string]: string } = {}
  Object.keys(parsed).forEach((key) => {
    defaultValues[key + '_key'] = key || ''
    defaultValues[key + '_value'] = parsed[key] || ''
    defaultValues[key + '_scope'] = APIVariableScopeEnum.PROJECT
    defaultValues[key + '_secret'] = ''
  })

  return defaultValues
}
