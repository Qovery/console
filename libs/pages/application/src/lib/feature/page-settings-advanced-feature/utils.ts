import { ApplicationAdvancedSettings } from 'qovery-typescript-axios'
import { ApplicationEntity } from '@console/shared/interfaces'

export function initFormValues(keys: string[], application: ApplicationEntity): { [key: string]: string } {
  const values: { [key: string]: string } = {}
  keys.forEach((key) => {
    if (application.advanced_settings?.current_settings) {
      values[key] =
        application.advanced_settings.current_settings[key as keyof ApplicationAdvancedSettings]?.toString() || ''
    }
  })

  return values
}
