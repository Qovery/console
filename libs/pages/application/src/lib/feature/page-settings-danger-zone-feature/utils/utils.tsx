import { UseFormRegister, UseFormUnregister } from 'react-hook-form/dist/types/form'

export const onAddStorage = (
  register: UseFormRegister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  const id = Math.trunc(Math.random() * 10000).toString()

  register('size_' + id, { value: '' })
  register('type_' + id, { value: '' })
  register('path_' + id, { value: '' })

  return [...keys, id]
}

export const onRemove = (
  key: string,
  unregister: UseFormUnregister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  unregister('size_' + key)
  unregister('type_' + key)
  unregister('path_' + key)

  return keys.filter((k) => k !== key)
}
