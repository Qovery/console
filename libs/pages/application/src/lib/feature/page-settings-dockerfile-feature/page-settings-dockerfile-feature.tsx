import { type LifecycleJobResponse } from 'qovery-typescript-axios'
import { type ReactNode } from 'react'
import { type UseFormReturn, useForm } from 'react-hook-form'
import { useParams } from 'react-router-dom'
import { match } from 'ts-pattern'
import { useCheckDockerfile, useLifecycleTemplate } from '@qovery/domains/environments/feature'
import {
  DockerfileSettings,
  type DockerfileSettingsData,
  useEditService,
  useService,
} from '@qovery/domains/services/feature'
import { SettingsHeading } from '@qovery/shared/console-shared'
import { isJobGitSource } from '@qovery/shared/enums'
import { ENVIRONMENT_LOGS_URL } from '@qovery/shared/routes'
import { Button } from '@qovery/shared/ui'
import { upperCaseFirstLetter } from '@qovery/shared/util-js'
import { TemplateIds } from '@qovery/shared/util-services'

function DockerfileSettingsFromTemplate({
  children,
  onSubmit,
  dockerfileForm,
  service,
}: {
  children: ReactNode
  onSubmit: () => void
  dockerfileForm: UseFormReturn<DockerfileSettingsData>
  service: LifecycleJobResponse
}) {
  const { data: template } = useLifecycleTemplate({
    environmentId: service.environment.id,
    templateId: TemplateIds[service.schedule.lifecycle_type as keyof typeof TemplateIds],
  })
  return (
    <DockerfileSettings
      methods={dockerfileForm}
      onSubmit={onSubmit}
      directSubmit
      defaultContent={template?.dockerfile}
      templateType={service.schedule.lifecycle_type}
    >
      {children}
    </DockerfileSettings>
  )
}

export function PageSettingsDockerfileFeature() {
  const { organizationId = '', projectId = '', environmentId = '', applicationId = '' } = useParams()

  const { data: service } = useService({ serviceId: applicationId, serviceType: 'JOB' })
  const { mutate: editService, isLoading: isLoadingEditService } = useEditService({
    environmentId,
    logsLink: ENVIRONMENT_LOGS_URL(organizationId, projectId, environmentId),
  })
  const { mutateAsync: mutateCheckDockerfile, isLoading: isLoadingCheckDockerfile } = useCheckDockerfile()

  const dockerfileForm = useForm<DockerfileSettingsData>({
    mode: 'onChange',
    defaultValues: {
      dockerfile_source:
        service?.job_type === 'LIFECYCLE' && isJobGitSource(service.source) && service.source.docker.dockerfile_raw
          ? 'DOCKERFILE_RAW'
          : 'GIT_REPOSITORY',
      dockerfile_path:
        service?.job_type === 'LIFECYCLE' && isJobGitSource(service.source)
          ? service.source.docker.dockerfile_path ?? 'Dockerfile'
          : undefined,
      dockerfile_raw:
        service?.job_type === 'LIFECYCLE' && isJobGitSource(service.source)
          ? service.source.docker.dockerfile_raw ?? ''
          : undefined,
    },
  })

  if (service?.job_type === 'LIFECYCLE' && isJobGitSource(service.source)) {
    const gitRepository = service.source.docker.git_repository
    const onSubmit = dockerfileForm.handleSubmit(async (data) => {
      if (data.dockerfile_raw) {
        editService({
          serviceId: applicationId,
          payload: {
            ...service,
            source: {
              docker: {
                git_repository: gitRepository,
                dockerfile_raw: data.dockerfile_raw,
                dockerfile_path: null,
              },
            },
          },
        })
      } else if (data.dockerfile_path) {
        try {
          await mutateCheckDockerfile({
            environmentId,
            dockerfileCheckRequest: {
              git_repository: {
                url: gitRepository?.url ?? '',
                root_path: gitRepository?.root_path,
                branch: gitRepository?.branch,
                git_token_id: gitRepository?.git_token_id,
              },
              dockerfile_path: data.dockerfile_path,
            },
          })
          editService({
            serviceId: applicationId,
            payload: {
              ...service,
              source: {
                docker: {
                  git_repository: gitRepository,
                  dockerfile_path: data.dockerfile_path,
                  dockerfile_raw: null,
                },
              },
            },
          })
        } catch (e: unknown) {
          dockerfileForm.setError('dockerfile_path', {
            type: 'custom',
            message: (e as Error).message ?? 'Dockerfile not found, please check the path and try again.',
          })
        }
      }
    })

    const watchDockerfilePath = dockerfileForm.watch('dockerfile_path')
    const watchDockerfileRaw = dockerfileForm.watch('dockerfile_raw')
    const watchDockerfileSource = dockerfileForm.watch('dockerfile_source')

    // HACK: Circonvent pitfall of formState.isValid issues, specially in test
    // https://github.com/react-hook-form/react-hook-form/issues/2755
    const isValid = watchDockerfileSource === 'DOCKERFILE_RAW' ? !!watchDockerfileRaw : !!watchDockerfilePath

    const DockerfileSettingsWrapper =
      service?.job_type === 'LIFECYCLE' && service.schedule.lifecycle_type !== 'GENERIC'
        ? ({ children }: { children: ReactNode }) => (
            <DockerfileSettingsFromTemplate onSubmit={onSubmit} dockerfileForm={dockerfileForm} service={service}>
              {children}
            </DockerfileSettingsFromTemplate>
          )
        : ({ children }: { children: ReactNode }) => (
            <DockerfileSettings methods={dockerfileForm} onSubmit={onSubmit} directSubmit>
              {children}
            </DockerfileSettings>
          )

    return (
      <div className="flex w-full flex-col justify-between">
        <div className="max-w-content-with-navigation-left p-8">
          <SettingsHeading
            title="Dockerfile"
            description={match(service.schedule.lifecycle_type)
              .with(
                'CLOUDFORMATION',
                (templateType) =>
                  `The Dockerfile allows to package your template and input into a container image with the right ${upperCaseFirstLetter(templateType)} CLI, inputs and commands to run. The Dockerfile can be stored in your git repository or on the Qovery control plane (Raw).`
              )
              .with(
                'TERRAFORM',
                (templateType) =>
                  `The Dockerfile allows to package your manifest and input into a container image with the right ${upperCaseFirstLetter(templateType)} CLI, inputs and commands to run. The Dockerfile can be stored in your git repository or on the Qovery control plane (Raw).`
              )
              .with(
                'GENERIC',
                undefined,
                () =>
                  `The Dockerfile allows to package your application with the right CLIs/Libraries and as well define the command to run during its execution. The Dockerfile can be stored in your git repository or on the Qovery control plane (Raw).`
              )
              .exhaustive()}
          />
          <DockerfileSettingsWrapper>
            <div className="flex justify-end">
              <Button
                type="submit"
                size="lg"
                disabled={isLoadingCheckDockerfile || !isValid}
                loading={isLoadingEditService}
              >
                Save
              </Button>
            </div>
          </DockerfileSettingsWrapper>
        </div>
      </div>
    )
  } else {
    // Could not happen as only lifecycle with git source have access to this page
    return null
  }
}

export default PageSettingsDockerfileFeature
