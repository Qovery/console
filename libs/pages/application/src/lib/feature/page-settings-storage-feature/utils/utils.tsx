import { UseFormRegister, UseFormUnregister } from 'react-hook-form/dist/types/form'
import { ApplicationStorageStorage } from 'qovery-typescript-axios'

export const addStorage = (register: UseFormRegister<{ [key: string]: string | number }>, keys: string[]): string[] => {
  const id = Math.trunc(Math.random() * 10000).toString()

  register('size_' + id, { value: '' })
  register('type_' + id, { value: '' })
  register('path_' + id, { value: '' })

  return [...keys, id]
}

export const onRemove = (
  key: string,
  unregister: UseFormUnregister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  unregister('size_' + key)
  unregister('type_' + key)
  unregister('path_' + key)

  return keys.filter((k) => k !== key)
}

export const initStorage = (
  register: UseFormRegister<{ [key: string]: string | number }>,
  storages: ApplicationStorageStorage[]
): string[] => {
  const keys: string[] = []

  storages.forEach((storage) => {
    register('size_' + storage.id, { value: storage.size })
    register('type_' + storage.id, { value: storage.type })
    register('path_' + storage.id, { value: storage.mount_point })

    keys.push(storage.id || '')
  })

  return keys
}
