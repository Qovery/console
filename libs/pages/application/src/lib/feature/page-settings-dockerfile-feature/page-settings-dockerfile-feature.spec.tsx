import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { renderWithProviders } from '@qovery/shared/util-tests'
import { PageSettingsDockerfileFeature } from './page-settings-dockerfile-feature'

describe('PageSettingsDockerfileFeature', () => {
  it('should match snapshot', () => {
    const { baseElement } = renderWithProviders(wrapWithReactHookForm(<PageSettingsDockerfileFeature />))
    expect(baseElement).toMatchSnapshot()
  })
})
