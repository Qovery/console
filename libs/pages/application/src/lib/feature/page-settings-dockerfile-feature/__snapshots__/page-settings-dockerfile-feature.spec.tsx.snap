// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSettingsDockerfileFeature should match snapshot 1`] = `
<body>
  <div />
</body>
`;
